src/mender-update/standalone/standalone.cpp

```diff
@@ -201,618 +222,435 @@ error::Error RemoveStateData(database::KeyValueDatabase &db) {
-		event_loop = make_shared<events::EventLoop>();
-		http_client =
-			make_shared<http::Client>(main_context.GetConfig().GetHttpClientConfig(), *event_loop);
-		auto reader = ReaderFromUrl(*event_loop, *http_client, src);
-		if (!reader) {
-			return {Result::FailedNothingDone, reader.error()};
-		}
-		artifact_reader = reader.value();
+error::Error StateMachine::SetStartStateFromStateData(const string &in_state) {
+	if (in_state == "ArtifactInstall") {
+		start_state_ = &artifact_install_enter_state_;
+	} else if (in_state == "ArtifactCommit") {
+		start_state_ = &artifact_commit_enter_state_;
+	} else if (in_state == "ArtifactCommit_Leave") {
+		start_state_ = &artifact_commit_leave_state_;
+	} else if (in_state == "Cleanup") {
+		start_state_ = &cleanup_state_;
+	} else if (in_state == "ArtifactRollback") {
+		start_state_ = &artifact_rollback_enter_state_;
+	} else if (in_state == "ArtifactFailure") {
+		start_state_ = &artifact_failure_enter_state_;
 	} else {
-		auto stream = io::OpenIfstream(src);
-		if (!stream) {
-			return {Result::FailedNothingDone, stream.error()};
-		}
-		auto file_stream = make_shared<ifstream>(std::move(stream.value()));
-		artifact_reader = make_shared<io::StreamReader>(file_stream);
-	}
-
-	string art_scripts_path = main_context.GetConfig().paths.GetArtScriptsPath();
-
-	// Clear the artifact scripts directory so we don't risk old scripts lingering.
-	auto err = path::DeleteRecursively(art_scripts_path);
-	if (err != error::NoError) {
-		return {Result::FailedNothingDone, err.WithContext("When preparing to parse artifact")};
+		return context::MakeError(
+			context::DatabaseValueError, "Invalid InState in database " + in_state);
 	}
 
-	artifact::config::ParserConfig config {
-		.artifact_scripts_filesystem_path = main_context.GetConfig().paths.GetArtScriptsPath(),
-		.artifact_scripts_version = 3,
-		.artifact_verify_keys = main_context.GetConfig().artifact_verify_keys,
-		.verify_signature = verify_signature,
-	};
-
-	auto exp_parser = artifact::Parse(*artifact_reader, config);
-	if (!exp_parser) {
-		return {Result::FailedNothingDone, exp_parser.error()};
-	}
-	auto &parser = exp_parser.value();
-
-	auto exp_header = artifact::View(parser, 0);
-	if (!exp_header) {
-		return {Result::FailedNothingDone, exp_header.error()};
-	}
-	auto &header = exp_header.value();
-
-	if (options != InstallOptions::NoStdout) {
-		cout << "Installing artifact..." << endl;
-	}
+	return error::NoError;
+}
 
-	if (header.header.payload_type == "") {
-		auto data = StateDataFromPayloadHeaderView(header);
-		return DoEmptyPayloadArtifact(main_context, data, options);
-	}
+error::Error StateMachine::SetStopAfterState(const string &state) {
+	using tf = common::state_machine::TransitionFlag;
+	using se = StateEvent;
+	auto &s = state_machine_;
 
-	update_module::UpdateModule update_module(main_context, header.header.payload_type);
+	// Replace transition in state machine in order to exit at given point.
```

Comment:
This is nice.

Reply:
Agree! This is one of the places where having a state machine really shines.